### ISSUE TITLE
403 aws access denied

### ISSUE THREAD

**Issue by emerak** [fog/fog#2701]
Hey, I've been trying to upload images to aws with carrierwave, currently I'm getting this error

```
Excon::Errors::Forbidden: Expected(200) <=> Actual(403 Forbidden)
  response => #<Excon::Response:0xd91cd0c @data={:body=>"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>34A7FEEC4AA9B144</RequestId><HostId>+pwgviyQ1kSSmtYrsoxu5i4oyqqwO2HSAbAeCCLqKa4IxWgWpPiPDMT3n+jOkmO9</HostId></Error>", :headers=>{"x-amz-request-id"=>"34A7FEEC4AA9B144", "x-amz-id-2"=>"+pwgviyQ1kSSmtYrsoxu5i4oyqqwO2HSAbAeCCLqKa4IxWgWpPiPDMT3n+jOkmO9", "Content-Type"=>"application/xml", "Transfer-Encoding"=>"", "Date"=>"Thu, 20 Feb 2014 00:55:57 GMT", "Connection"=>"close", "Server"=>"AmazonS3"}, :status=>403, :remote_ip=>"207.171.163.152"}, @body="<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>34A7FEEC4AA9B144</RequestId><HostId>+pwgviyQ1kSSmtYrsoxu5i4oyqqwO2HSAbAeCCLqKa4IxWgWpPiPDMT3n+jOkmO9</HostId></Error>", @headers={"x-amz-request-id"=>"34A7FEEC4AA9B144", "x-amz-id-2"=>"+pwgviyQ1kSSmtYrsoxu5i4oyqqwO2HSAbAeCCLqKa4IxWgWpPiPDMT3n+jOkmO9", "Content-Type"=>"application/xml", "Transfer-Encoding"=>"", "Date"=>"Thu, 20 Feb 2014 00:55:57 GMT", "Connection"=>"close", "Server"=>"AmazonS3"}, @status=403, @remote_ip="207.171.163.152">
from /home/a/.rvm/gems/ruby-2.0.0-p353@i/gems/excon-0.31.0/lib/excon/middlewares/expects.rb:6:in `response_call'
```

I tried updating gems and I'm in the latest version, I also check other posts about that error but nothing seems to work, do I missing something?

**Comment by ideaoforder** [comment]
+1 - Exact same issue. Works fine in production, but not from our local dev server. I'm on localhost--maybe that's part of the problem? I'm using 1.27.0.

**Comment by ploomans** [comment]
I had the same issue but for me both production and development didn't work when I used the latest fog version together with rails 4.2.0

I had an old code base for another project with rails 4.1.1 and fog 1.3.1 where the same credentials did work. I just solved it for the moment by using the same 4.1.1 and fog 1.3.1 gem version for my new project...

**Comment by josetapadas** [comment]
Confim that it still happens on fog-1.32.0 with rails-4.2.

Somehow, when trying to write to a S3 Bucket via an AWS machine it retrieves "Access Denied", even though thru our local dev machines, with the same keys, it worked.

My solution was switching to ```carrierwave-aws``` since downgrading rails was out of the question.

**Comment by geemus** [comment]
@josetapadas thanks for the info, glad to hear you found a workaround.

**Comment by josetapadas** [comment]
@geemus after further investigation (even with other rails apps that were using fog as well) there seems to be an (unconfirmed) pattern:

It only happens when the app is trying to authenticate to AmazonS3 from an AWS machine. Does this make any sense to you guys?

**Comment by geemus** [comment]
Nothing jumps out at me as an obvious culprit from that, at least if the same credentials are used in both places, I would expect the same behavior from both.

**Comment by raldred** [comment]
I can confirm fog 1-32.1 has the same issue. My uploads are created and uploaded to S3 fine but fog still get's stuck with access denied.
Switched to carrierwave-aws and all is well.

**Comment by geemus** [comment]
I think in some cases it may also try to lookup permissions and/or presence of objects, so maybe `GetObjectAcl` and/or `HeadObject` if I remember right. So if those are not enabled, I could see that also causing issue.

**Comment by raldred** [comment]
I resolved my particular issue here. It requires DeleteObject too if it's replacing and item with a different filename.

**Comment by geemus** [comment]
@raldred ah, yeah that would do it. Thanks for the update!

**Comment by palfvin** [comment]
FWIW, I'm having the same issue with Fog 1.34 and Rails 4.2 using an AWS-compatible system (RiakCS). Same code and credentials work fine with Rails 4.1 and and earlier version of Fog as well as with the s3cmd command line tool. Thanks to others who have posted their experience and workarounds as well. Will likely try and drill down tomorrow in hopes of finding out what's going on.

**Comment by geemus** [comment]
@palfvin thanks for the heads up/details. Let me know if there is anything I can do to help.

**Comment by szajbus** [comment]
Had similar issue which I solved by setting bucket policy

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "S3:*",
            "Resource": "arn:aws:s3:::bucketname/*",
            "Condition": {}
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket",
                "s3:GetBucketLocation"
            ],
            "Resource": "arn:aws:s3:::bucketname",
            "Condition": {}
        }
    ]
}
```

**Comment by davidlesches** [comment]
@szajbus  thanks so much, that fixed it for me.

**Comment by raldred** [comment]
I don't think amazon could have created a more overly engineered permission system if they tried.

**Comment by abuisman** [comment]
@szajbus Aren't you enabling everyone in the world to do anything to that bucket with that policy?

**Comment by bitless** [comment]
@abuisman this is the bucket access policy attached to the fog iam user. to the op issue, in 1.40 we found that GetBucketLocation grant was necessary to allow the user to connect with fog console.

**Comment by sshakil** [comment]
I was also experiencing this issue, where uploads using fog (not aws specific).  I had tried both from my local machine and heroku to s3, with a 403 in both cases.  AWS CLI on my machine to s3 was fine.  Then I tried fine tuning the permissions, but wasn't able to get that working, until I tried the permission set given by szajbus on Oct 21, 2016.  That did the trick.  Time to continue fiddling with permissions.

**Comment by ninjarails** [comment]
Exact same issue

**Comment by timparker87** [comment]
@szajbus thank you, your solution worked for me. I had to additionally specify a `"Resource"` on the bucket policy, which in my case was the IAM account ARN (more info in the [docs](https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-bucket-user-policy-specifying-principal-intro.html) for anyone else who it may help)

And thank you @raldred for the heads-up on the DeleteObject permission!

**Comment by caseyli** [comment]
Hey everybody, I know this issue is closed, but I did run into something very similar to this and it turns out it was due to the default settings on the Amazon S3 Buckets. I think they must have changed this somewhere along the line.

Anyway, my basic fix was to NOT block public acl objects by going to the "Permissions" tab on your bucket.

I've detailed my full fix here: https://www.bitesite.ca/blog/fixing-rails-carrierwave-amazon-s3-403-forbidden-error

**Comment by geemus** [comment]
@caseyli thanks for taking time to share.

**Comment by pvin** [comment]
This discussion is very useful. We faced similar issue. 

[link](https://stackoverflow.com/questions/60483527/issue-in-aws-s3-bucket-with-x-amz-server-side-encryption-policy-not-allowing-rai/60490137#60490137)
 
**Issue in aws s3 bucket with x-amz-server-side-encryption policy not allowing rails5 to upload images but allows in rails4**

Fixed it by downgrade gem from carrierwave-2.0.2 to carrierwave-1.3.1 .**Could not fix the issue in 2.0.2 version**

This may help anyone...
